File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        int n = s.length();
        //index
        Stack<Integer> bucket = new Stack<>();
        char charArr[] = s.toCharArray();
        
        for(int i=0; i<n; i++){
            if(charArr[i] == '('){
                bucket.add(i);
            }
            else if(charArr[i] == ')'){
                if(bucket.isEmpty()){
                    charArr[i] = '*'; //invalid
                }
                else{
                    bucket.pop();
                }
            }
        }
        while(!bucket.isEmpty()){
            charArr[bucket.peek()] = '*'; //invalid
            bucket.pop();
        }
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<n; i++){
            if(charArr[i] != '*'){
                sb.append(charArr[i]);
            }
        }
        return sb.toString();
    }
}

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        int n = s.size();
        //index
        stack<int> bucket;
        for(int i=0; i<n; i++){
            if(s[i] == '('){
                bucket.push(i);
            }
            else if(s[i] == ')'){
                if(bucket.empty()){
                    s[i] = '*'; //invalid
                }
                else{
                    bucket.pop();
                }
            }
        }
        
        while(!bucket.empty()){
            s[bucket.top()] = '*'; //invalid
            bucket.pop();
        }
        s.erase(remove(s.begin(), s.end(), '*'), s.end());
        return s;
    }
};